windows-icon-handler/check-icon-handler/CheckIconHandler.cpp

```diff
@@ -0,0 +1,64 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 (eventually AGPL-3.0) 2016-present Scille SAS
+
+// CheckIconHandler.cpp : Implementation of CCheckIconHandler
+
+#include "pch.h"
+#include "CheckIconHandler.h"
+
+#include <array>
+#include <memory>
+
+
+// CCheckIconHandler
+
+HRESULT __stdcall CCheckIconHandler::IsMemberOf(LPCWSTR pwszPath, DWORD dwAttrib)
+{
+    constexpr std::array<wchar_t, 26> DRIVES = {
+        'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M',
+        'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z'
+    };
+    int drive_number = PathGetDriveNumberW(pwszPath);
+
+    //// Parsec supports drive from H to Z (included). If the drive letter is not
+    //// one of those, we can return FALSE immediately
+    //// A is 0, Z is 25, our drive should be between 7 and 25 included
+
+    if (drive_number == -1 || drive_number < 7 || PathIsRootW(pwszPath))
+        return S_FALSE;
+
+    // Parsec write a registry key to set the icon for its workspaces.
+    // We check if the key is present for the current drive.
+    auto reg_key = std::make_unique<wchar_t[]>(MAX_PATH * 32);
+    swprintf(
+        reg_key.get(),
+        MAX_PATH * 32,
+        L"Software\\Classes\\Applications\\Explorer.exe\\Drives\\%c\\DefaultIcon",
+        DRIVES[drive_number]
+    );
+    HKEY key = nullptr;
+
+    if (RegOpenKeyEx(HKEY_CURRENT_USER, reg_key.get(), 0, KEY_READ, &key) != ERROR_SUCCESS)
+        return S_FALSE;
+
+    return S_OK;
+}
+
+HRESULT __stdcall CCheckIconHandler::GetOverlayInfo(LPWSTR pwszIconFile, int cchMax, int* pIndex, DWORD* pdwFlags)
+{
+    // TODO: set standard icon path or load it once in memory?
+    const LPWSTR path = L"C:\\Users\\Corentin\\Documents\\icon_handler\\vs\\check.ico";
+    //const LPWSTR path = L"E:\\projects\\icon_handler\\check.ico";
+
+    // TODO: check if this file has its `__parsecinfo__`, parse its json content, set the appropriate icon if needed
```

Comment:
Wrong todo: the icon is always the same (and should be included as a resource in the DLL and loaded from there). `IsMemberOf` should be updated to check for the `.__parsec_entry_info__` file.

Reply:
Done.